processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -627,98 +624,96 @@ IncrementalIndexRowResult toIncrementalIndexRow(InputRow row)
-        if (!capabilities.hasMultipleValues() &&
-            dimsKey != null &&
-            handler.getLengthOfEncodedKeyComponent(dimsKey) > 1) {
-          capabilities.setHasMultipleValues(true);
+        DimensionHandler handler = DimensionHandlerUtils.getHandlerFromCapabilities(dimension, capabilities, null);
+        if (dimensionData == null) {
+          dimensionData = prevDimensionData.clone();
         }
-
-        if (wasNewDim) {
-          if (overflow == null) {
-            overflow = new ArrayList<>();
-          }
-          overflow.add(dimsKey);
-        } else if (desc.getIndex() > dims.length || dims[desc.getIndex()] != null) {
-          /*
-           * index > dims.length requires that we saw this dimension and added it to the dimensionOrder map,
-           * otherwise index is null. Since dims is initialized based on the size of dimensionOrder on each call to add,
-           * it must have been added to dimensionOrder during this InputRow.
-           *
-           * if we found an index for this dimension it means we've seen it already. If !(index > dims.length) then
-           * we saw it on a previous input row (this its safe to index into dims). If we found a value in
-           * the dims array for this index, it means we have seen this dimension already on this input row.
-           */
-          throw new ISE("Dimension[%s] occurred more than once in InputRow", dimension);
-        } else {
-          dims[desc.getIndex()] = dimsKey;
+        desc = dimensionData.addNewDimension(dimension, capabilities, handler);
+      }
+      DimensionHandler handler = desc.getHandler();
+      DimensionIndexer indexer = desc.getIndexer();
+      Object dimsKey = null;
+      try {
+        dimsKey = indexer.processRowValsToUnsortedEncodedKeyComponent(
+            row.getRaw(dimension),
+            true
+        );
+      }
+      catch (ParseException pe) {
+        parseExceptionMessages.add(pe.getMessage());
+      }
+      dimsKeySize += indexer.estimateEncodedKeyComponentSize(dimsKey);
+      // Set column capabilities as data is coming in
+      if (!capabilities.hasMultipleValues() &&
+          dimsKey != null &&
+          handler.getLengthOfEncodedKeyComponent(dimsKey) > 1) {
+        if (dimensionData == null) {
+          dimensionData = prevDimensionData.clone();
         }
+        capabilities = dimensionData.getDimensionCapabilities(dimension);
+        capabilities.setHasMultipleValues(true);
       }
+      rowDimKeys.put(dimension, dimsKey);
     }
 
-    if (overflow != null) {
-      // Merge overflow and non-overflow
-      Object[] newDims = new Object[dims.length + overflow.size()];
-      System.arraycopy(dims, 0, newDims, 0, dims.length);
-      for (int i = 0; i < overflow.size(); ++i) {
-        newDims[dims.length + i] = overflow.get(i);
+
+    if (dimensionData != null) {
+      while (!dimensions.compareAndSet(prevDimensionData, dimensionData)) {
```

Comment:
I think it would be simpler to do double-checked locking here than a CAS loop.

Reply:
From what I gather, there's not a lot of contention in updating dimension data. The motivation to use CAS rather than locking was to save thread state changes, which may result in more costly context switching.